### PR TITLE
qa: add test to mangle lost+found directory object and ensure safety

### PR DESCRIPTION
**Pull request by vshankar**
Earlier ceph versions didn't allow the lost+found directory to be removed
and nor the entries inside it. Users are recommended to fail the filesystem
and remove the directory object using rados cli commands. Therefore, include
this step as part of our testing.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by vshankar** [comment]
https://pulpito.ceph.com/vshankar-2023-10-11_12:16:08-fs-main-testing-default-smithi/7421163/

**Review by chrisphoffman — COMMENTED** [review]
Do you have any more information on the motivation for the test? For instance: documentation, a tracker, or ceph mailing list etc?

**Comment by dparmar18** [comment]
> Do you have any more information on the motivation for the test? For instance: documentation, a tracker, or ceph mailing list etc?

It should be this https://tracker.ceph.com/issues/59569, there were some conversation about this on upstream mailing list too https://lists.ceph.io/hyperkitty/list/ceph-users@ceph.io/message/FIBTZF7IAZVMHU2YFMP4WCH5SGCJ3S5M/

**Comment by vshankar** [comment]
> > Do you have any more information on the motivation for the test? For instance: documentation, a tracker, or ceph mailing list etc?

Good question and I should have been a bit more elaborate on this - we recommend these steps as a workaround for the issue @dparmar18 mentioned, so I added a test to verify it.

File: qa/tasks/cephfs/test_data_scan.py [hunk 1]
```diff
@@ -190,6 +200,23 @@ def validate(self):
...
+    def flush(self):
+        # Never flush metadata, so backtrace won't be written
+        pass
+
+    def mangle(self):
+        self._filesystem.rados(["-p", self._filesystem.get_metadata_pool_name(), "rm", "4.00000000"])
+        self._filesystem.rados(["-p", self._filesystem.get_metadata_pool_name(), "rmomapkey", "1.00000000", "lost+found_head"])
+
+    def validate(self):
+        # The inode should be linked into lost+found because we had no path for it
```
Comment (batrick) [inline qa/tasks/cephfs/test_data_scan.py:218]: ? This test removes lost+found. So this comment doesn't make sense to me.

**Review by batrick — CHANGES_REQUESTED** [review]

File: qa/tasks/cephfs/test_data_scan.py (hunk 1 above)
Comment (vshankar) [reply qa/tasks/cephfs/test_data_scan.py:218]: Comment is stale since the most of this test was copied from an existing test. Will remove.

**Comment by vshankar** [comment]
@mchangir @rishabh-d-dave integration test ping...

**Review by kotreshhr — APPROVED** [review]

**Comment by vshankar** [comment]
> @mchangir @rishabh-d-dave integration test ping...

ping @mchangir @rishabh-d-dave

File: qa/tasks/cephfs/test_data_scan.py [hunk 2]
```diff
@@ -190,6 +200,23 @@ def validate(self):
...
+    def flush(self):
+        # Never flush metadata, so backtrace won't be written
+        pass
+
+    def mangle(self):
+        self._filesystem.rados(["-p", self._filesystem.get_metadata_pool_name(), "rm", "4.00000000"])
+        self._filesystem.rados(["-p", self._filesystem.get_metadata_pool_name(), "rmomapkey", "1.00000000", "lost+found_head"])
+
+    def validate(self):
+        # The dir should be gine since we manually removed it
```
Comment (dparmar18) [inline qa/tasks/cephfs/test_data_scan.py:218]: nit: s/gine/gone

Comment (vshankar) [reply qa/tasks/cephfs/test_data_scan.py]: Done.

File: qa/tasks/cephfs/test_data_scan.py [hunk 3]
```diff
@@ -101,6 +101,16 @@ def scrub(self):
         self.assert_equal(out_json["return_code"], 0)
         self.assert_equal(self._filesystem.wait_until_scrub_complete(tag=out_json["scrub_tag"]), True)
 
+    def mangle(self):
+        """
+        Gives an opportunity to fiddle with metadata objects before bringing back
+        the MDSs online. This is used in testing the lost+found case (when recovering
+        a file without a backtrace) to verify if lost+found directory object can be
+        removed via RADOS operation and the file system can be continued to be used
+        as expected.
```
Comment (batrick) [inline qa/tasks/cephfs/test_data_scan.py:110]: Sorry, in which situation are we recommending users to delete (using `rados`) the lost+found directory? Your commit message says:

> Users are recommended to fail the filesystem and remove the directory object using rados cli commands.

I thought the issue to test was that a user could unlink or rename files normally inside the lost+found directory?

**Review by batrick — CHANGES_REQUESTED** [review]

File: qa/tasks/cephfs/test_data_scan.py (hunk 3 above)
Comment (vshankar) [reply qa/tasks/cephfs/test_data_scan.py:110]: > Sorry, in which situation are we recommending users to delete (using `rados`) the lost+found directory? Your commit message says:
> 
> > Users are recommended to fail the filesystem and remove the directory object using rados cli commands.
> 
> I thought the issue to test was that a user could unlink or rename files normally inside the lost+found directory?

This was broken and was fixed with commit 241b9396dbc6d506cef297bdb89ab07fe3a7ce02. So, users running older versions would have to cleanup via removing omap keys. That's what this test verifies.

**Comment by vshankar** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Main#29-Jan-2024

**Pull request merged**